src/catalog/database_catalog.cpp

```diff
@@ -0,0 +1,1643 @@
+  auto dsrc_entry =
+      reinterpret_cast<storage::VarlenEntry *>(redo->Delta()->AccessForceNotNull(table_pm[ADSRC_COL_OID]));
+  *oid_entry = !col.GetOid();
+  *relid_entry = class_oid;
+  const storage::VarlenEntry name_varlen = postgres::AttributeHelper::MakeNameVarlen<Column>(col);
+  *name_entry = name_varlen;
+  *type_entry = col.GetType();
+  // TODO(Amadou): Figure out what really goes here for varlen. Unclear if it's attribute size (16) or varlen length
+  *len_entry = (col.GetType() == type::TypeId::VARCHAR || col.GetType() == type::TypeId::VARBINARY)
+                   ? col.GetMaxVarlenSize()
+                   : col.GetAttrSize();
+  *notnull_entry = !col.GetNullable();
+  *dbin_entry = reinterpret_cast<intptr_t>(default_val);
+  storage::VarlenEntry dsrc_varlen = storage::StorageUtil::CreateVarlen(default_val->ToJson().dump());
+  *dsrc_entry = dsrc_varlen;
+  // Finally, insert into the table to get the tuple slot
+  const auto tupleslot = columns_->Insert(txn, redo);
+
+  // Step 2: Insert into name index
+  const auto name_pri = columns_name_index_->GetProjectedRowInitializer();
+  // Create a buffer large enough for all columns
+  auto *const buffer = common::AllocationUtil::AllocateAligned(name_pri.ProjectedRowSize());
+  auto *pr = name_pri.InitializeRow(buffer);
+  // Write the attributes in the ProjectedRow. We know the offsets without the map because of the ordering of attribute
+  // sizes
+  *(reinterpret_cast<storage::VarlenEntry *>(pr->AccessForceNotNull(0))) = name_varlen;
+  *(reinterpret_cast<uint32_t *>(pr->AccessForceNotNull(1))) = class_oid;
+
+  if (!columns_name_index_->InsertUnique(txn, *pr, tupleslot)) {
+    // There was a name conflict and we need to abort.  Free the buffer and return false to indicate failure
+    delete[] buffer;
+
+    // Clean up the varlen's buffer in the case it wasn't inlined.
+    if (!name_varlen.IsInlined()) {
+      delete[] name_varlen.Content();
+    }
+
+    return false;
+  }
+
+  // Clean up the varlen's buffer in the case it wasn't inlined.
+  if (!name_varlen.IsInlined()) {
+    delete[] name_varlen.Content();
+  }
+
+  // Step 3: Insert into oid index
+  const auto oid_pri = columns_oid_index_->GetProjectedRowInitializer();
+  const auto oid_pm = columns_oid_index_->GetKeyOidToOffsetMap();
+  pr = oid_pri.InitializeRow(buffer);
+  // Write the attributes in the ProjectedRow. These hardcoded indexkeycol_oids come from
+  // Builder::GetColumnOidIndexSchema()
+  *(reinterpret_cast<uint32_t *>(pr->AccessForceNotNull(oid_pm.at(indexkeycol_oid_t(2))))) = col.GetOid();
+  *(reinterpret_cast<uint32_t *>(pr->AccessForceNotNull(oid_pm.at(indexkeycol_oid_t(1))))) = class_oid;
+
+  if (!columns_oid_index_->InsertUnique(txn, *pr, tupleslot)) {
+    // There was an oid conflict and we need to abort.  Free the buffer and return false to indicate failure
+    // TODO(Matt): Is this is possible at this point, or should we be asserting the insert result like the oid index on
```

Comment:
See TODO.

Reply:
I would lean towards this being an assert and force column OID reasoning to the class (i.e. index and table) logic.  That's where we should be doing OID assignments and management since we need to track it in order to support multi-versioned schemas in the future.